### ISSUE TITLE
Reevaluate flushing unschedulable pods into activeQ

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#87850]
/sig scheduling

#72122 suggested the addition of flushing unschedulable pods into the active queue, to mitigate against unknown race conditions that would keep pods in the unschedulable queue forever. The scenario I can think of is that an event that makes the pod more schedulable comes while we are still in its scheduling cycle. But this should have been addressed with #73078.

In any case, as seen in #86373, flushing indistinctively can produce starvation if there are too many unschedulable pods with high-priority, or if scheduling takes too long due to the use of advanced features (like extenders).

So I suggest any of these 2 options:

- [Extreme] Get rid of the flush, since we don't have evidence that it's actually helping more than #73078.
- Limit the number of pods we flush back. We would always flush the oldest ones. Either a (configurable?) hard limit (always flush X) or a dynamic limit (stop once the activeQ is past a threshold). This should be low risk and easy to do.

**Comment by alculquicondor** [comment]
cc @ahg-g @Huang-Wei

**Comment by ahg-g** [comment]
To help make an informed decision, we need a simulation test to evaluate how big of a problem this actually is and in what scenarios. 

The test should be similar to [TestBackOffFlow](https://github.com/kubernetes/kubernetes/blob/d90dd93855ab6c3ccda73dfbd3d19c81f56abca5/pkg/scheduler/internal/queue/scheduling_queue_test.go#L1519), and one of its parameters should be the schedule's performance (expressed in units of time used to advance the clock for each scheduled pod).

**Comment by Huang-Wei** [comment]
Agree to have the simulation test ready first.

> [Extreme] Get rid of the flush, since we don't have evidence that it's actually helping more than #73078.

Technically I'm leaning towards this solution. But we should be very cautious on this and need thorough tests to catch any possible corner cases (which is hard...). I raised this is because sometimes technically, it's a better design, but hard to carry out. I recalled when we exercised the idea of "only put pods back into activeQ when the cluster becomes more schedulable", we did trigger a couple of regressions which was reported by end-users.

**Comment by alculquicondor** [comment]
/help

This is not necessarily a good first issue, as it requires experience in the code base and good understanding of the queueing flow. I don't have time to work in this now, but I can provide guidance to a relatively experienced contributor.

**Comment by mateuszlitwin** [comment]
I am concerned that if you decide to remove period flush then pods will remain in the unschedulableQ for long time. This can happen in custom scheduling plugin where there are events that should move pod to activeQ, but are not tracked by the default scheduler. It was mentioned in #77746 and recently by me in #87738.

**Comment by alculquicondor** [comment]
That should be solved separately according to use cases. That is to say, those kind of features shouldn't depend on the periodic flush.

**Comment by alculquicondor** [comment]
@denkensk are you still working on the simulation?

**Comment by denkensk** [comment]
> @denkensk are you still working on the simulation?

Sorry for the slow response  @alculquicondor  I will send a pr as soon as possible

**Comment by alculquicondor** [comment]
Any progress on this?

**Comment by fengzixu** [comment]
@denkensk Hi. Are you still working on this issue?

**Comment by alculquicondor** [comment]
cc @adtac

**Comment by alculquicondor** [comment]
It's complementary. The flush still happens even if there are no events.

**Comment by Huang-Wei** [comment]
Flushing behavior may still be necessary as some resources not managed by k8s (e.g., hostPort) may change over time, and that may make the current unschedulable Pod schedulable.

**Comment by alculquicondor** [comment]
I don't follow. Isn't that part of the PodSpec?

**Comment by Huang-Wei** [comment]
> I don't follow. Isn't that part of the PodSpec?

I mean some hostPorts may be occupied by system software that is not managed by k8s. If we don't do regular flushing at all, the unschedulable Pod won't get a chance to be re-schedule even if the system resource gets released.

But this may not be a good example as the Pod would have been evicted by kubelet upon hostPort conflicts instead of being held in scheduler queue. A more practical example is some resources (CRD or Cloud Infra resources) are considered upon scheduling, but not considered upon re-queuing. In that case, disabling flushing may keep them forever in a unschedulable state.

**Comment by alculquicondor** [comment]
Sounds good. Maybe we can remove the flush once we provide the proper extension points for plugins to trigger re-queuing.

**Comment by alculquicondor** [comment]
/remove-lifecycle rotten

@Huang-Wei @denkensk, after all the improvements we have made with event handling, what do you think of (at least) increasing the flush period.

The flush is harmful during cluster autoscaler scale ups, because a Pod is retried more than needed, potentially increasing it's backoff. Then newer pods could end up being scheduled starving older pods.

**Comment by alculquicondor** [comment]
5min sgtm, can you send a PR?

**Comment by denkensk** [comment]
Sure

**Comment by ahg-g** [comment]
+1

**Comment by yuanchen8911** [comment]
> 5min sgtm, can you send a PR?

5 min as default, but make it configurable?

**Comment by alculquicondor** [comment]
Correct, we want to eventually eliminate it. If we don't see bug reports, we can probably do so in a couple of releases.

**Comment by Huang-Wei** [comment]
Increasing the internal is a good interim step. However, to smooth the elimination process, I think there are some gaps to better support the decision: like metrics and(or) events to show pods number/names moved by the enforced periodic flush. We may want to evaluate the data points how these metrics/events look like on a prod cluster.

**Comment by alculquicondor** [comment]
There will always be pods like that are moved by the flush (for example, in a steady cluster where no pods finished and no new nodes are added).

Perhaps something more useful is whether a pod got scheduled after getting put in the queue due to a flush. And then we should include which reason made it unschedulable in the past, so we can possibly identify which event we missed. Any other information that would be useful?

**Comment by Huang-Wei** [comment]
> And then we should include which reason made it unschedulable in the past

Yes, that basically what I meant above as metrics/events: if we observed that the metric showing the # of pods being moved by flush is 0; or if it's a small number, and by looking into the events/logs, we verify that the number is as expected, they are good signals to retire the flush interval.

We don't have "reason" in the `QueuePodInfo`, but `UnschedulablePlugins` is available and that should suffice to compose the event/log.

> Any other information that would be useful?

Use QueuedPodInfo's {Attempts, UnschedulablePlugins} as tuples to compare on each flushing ticks should be useful.

**Comment by alculquicondor** [comment]
> if we observed that the metric showing the # of pods being moved by flush is 0; or if it's a small number, and by looking into the events/logs, we verify that the number is as expected, they are good signals to retire the flush interval.

I think verifying that the number is as expected is terribly difficult unless you are fully familiar with the pods in a cluster, which is not realistic when there are cluster administrators and support engineers from cloud providers.

**Comment by ahg-g** [comment]
logs/events are not practical.

I think a metric tracking the number of scheduled pods broken down by both `UnschedulablePlugins` (we can just use any one) and `ClusterEvent` triggering the pod to be moved to the active queue when it got scheduled would be useful, this will allow us to compare across all events, and the flush event should be extremely low relatively if indeed flushing is not needed (there will be cases were flushing will be the cause because of the following race: a flush event triggers right before the event that would have caused the pod to be moved back to the queue and get scheduled).

If we gradually increase the flush period across few versions, then we should see that percentage trend down to zero, which I think will be a good signal to eliminate it eventually, the problem though is that we can't change the default value with bumping the ComponentConfig version, so not sure how to achieve that while respecting CC versioning.

**Comment by Huang-Wei** [comment]
> If we gradually increase the flush period across few versions, then we should see that percentage trend down to zero, which I think will be a good signal to eliminate it eventually

Reasonable metric(s) would suffice to tell if we can retire the flushing interval, but only on a happy path. If it turns out to be undesired number, we need extra info like logs/events to pinpoint the problem. I'd still suggest to go with metric and logs (events can be emitted due to extra costs). 

> the problem though is that we can't change the default value with bumping the ComponentConfig version, so not sure how to achieve that while respecting CC versioning.

Although a bit hacky, maybe plumb an environmental variable? so we can make it customized while also not bothering with CC.

(It's not that uncommon in k8s codebase: like [HTTP2_READ_IDLE_TIMEOUT_SECONDS](https://github.com/kubernetes/kubernetes/blob/8c69e5d25bfdbd266b384db29493bc9bad60f092/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L149) and other occurrences of `os.Getenv`)

**Comment by denkensk** [comment]
> I think a metric tracking the number of scheduled pods broken down by both UnschedulablePlugins (we can just use any one) and ClusterEvent triggering the pod to be moved to the active queue when it got scheduled would be useful

I suppose that maybe the metric with multi-dim labels will help us to evaluate if we can retire the flushing interval and identify which event we missed.
- (1 dim): The reason triggers the pod to be moved to the active queue, It may includes flushing event, node/add, pod/delete.... (It can be divided into only two categories: flush/event). This will help us to evaluate if we can retire the flushing interval depending on the number of flush.
- (2 dim): `UnschedulablePlugins`.  This will help us to identify which event we missed through `group by` use the label `enqueueReason: flush`. 

> Although a bit hacky, maybe plumb an environmental variable? so we can make it customized while also not bothering with CC.

If it is just an experimental parameter to verify the feature, I think environment variables are fine.

**Comment by ahg-g** [comment]
Logging is great for debugging, so that sgtm; but it is not practical to spot trends. so I guess we are on the same page.

Not everyone uses the scripts in the repo to start k8s. I think we can just have this as a CC, and cloud providers should tune it at the pace they see fit.

**Comment by ahg-g** [comment]
but that is not safe, we need an easy way to change it back if increasing it caused problems.

**Comment by alculquicondor** [comment]
>  It can be divided into only two categories: flush/event

I think we can simply consider the flush yet another type of event. If we do this, the metric itself can serve us long term and we don't need to retire it once we remove the flush.

> but that is not safe, we need an easy way to change it back if increasing it caused problems.

Ah, good point, then it should either be a command line flag and/or environment variable that gets released with a deprecated notice from the beginning.

**Comment by ahg-g** [comment]
> I think we can simply consider the flush yet another type of event. If we do this, the metric itself can serve us long term and we don't need to retire it once we remove the flush.

it is already one: https://github.com/kubernetes/kubernetes/blob/0669da445fa8c1ae07c15c0827f0e83da11cbe58/pkg/scheduler/internal/queue/events.go#L90

> Ah, good point, then it should either be a command line flag and/or environment variable that gets released with a deprecated notice from the beginning.

a deprecated flag sounds like a good option

**Comment by Huang-Wei** [comment]
> I think we can simply consider the flush yet another type of event

Yes, it already is:

https://github.com/kubernetes/kubernetes/blob/0669da445fa8c1ae07c15c0827f0e83da11cbe58/pkg/scheduler/internal/queue/events.go#L89-L90

**Comment by alculquicondor** [comment]
Let's remove the timeout in 1.27? We set the timeout to 5min in 1.24.

**Comment by ahg-g** [comment]
I wonder if we should keep it, expose it via component config, and set it to something very high by default (like 24 hours). The reason: this perhaps can be useful as a mitigation knob.

**Comment by denkensk** [comment]
> I wonder if we should keep it, expose it via component config, and set it to something very high by default (like 24 hours). 

It has been exposed by the command line parameter, do you suggest to expose it in `KubeSchedulerConfiguration` instead? 

> The reason: this perhaps can be useful as a mitigation knob.

I think it's a more conservative and safer option. And we can change the default time to a long time (12 hours or 24hours).

**Comment by ahg-g** [comment]
> It has been exposed by the command line parameter, do you suggest to expose it in KubeSchedulerConfiguration instead?

Yes, we don't want to rely on flags for scheduler configuration.

**Comment by denkensk** [comment]
> Yes, we don't want to rely on flags for scheduler configuration.

Ok. Got it. I will move it to the `KubeSchedulerConfiguration` in  1.27.

**Comment by sanposhiho** [comment]
As discussed here, we concluded that we make the queueing hint focus on the event that actually makes Pod schedulable.
https://github.com/kubernetes/kubernetes/pull/119396#discussion_r1324024982

That means that, in worst-case scenario, if we have a bug like https://github.com/kubernetes/kubernetes/pull/120334, Pods may be stacked in the unschedulable Pod pool forever. So, let's wait removing flushing while we confirm that no such bug report after all QueueingHint is implemented https://github.com/kubernetes/kubernetes/issues/118893.

**Comment by sanposhiho** [comment]
preCheck is removed successfully when QHint is disabled.

Another corner case that we need a flush and hence need to solve before the complete removal: https://github.com/kubernetes/kubernetes/issues/126700
